source/extensions/filters/network/proxy_protocol/proxy_protocol.cc

```diff
@@ -0,0 +1,185 @@
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_SIGNATURE_LEN;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_TRANSPORT_STREAM;
+using Envoy::Extensions::Filters::Common::ProxyProtocol::PROXY_PROTO_V2_VERSION;
+using envoy::extensions::filters::network::proxy_protocol::v3::ProxyProtocol_Version;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ProxyProtocol {
+
+Filter::Filter(const envoy::extensions::filters::network::proxy_protocol::v3::ProxyProtocol& config)
+    : config_(config) {}
+
+Network::FilterStatus Filter::onWrite(Buffer::Instance&, bool) {
+  if (!sent_header_) {
+    injectHeader();
+    sent_header_ = true;
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void Filter::injectHeader() {
+  if (config_.version() == ProxyProtocol_Version::ProxyProtocol_Version_V1) {
+    injectHeaderV1();
+  } else if (config_.version() == ProxyProtocol_Version::ProxyProtocol_Version_V2) {
+    injectHeaderV2();
+  }
+}
+
+void Filter::injectHeaderV1() {
+  std::ostringstream stream;
+  stream << PROXY_PROTO_V1_SIGNATURE;
+  // Default to local addresses
+  std::string src_address = write_callbacks_->connection().localAddress()->ip()->addressAsString();
+  std::string dst_address = write_callbacks_->connection().remoteAddress()->ip()->addressAsString();
+  auto src_port = write_callbacks_->connection().localAddress()->ip()->port();
+  auto dstPort = write_callbacks_->connection().remoteAddress()->ip()->port();
+  auto ip_version = write_callbacks_->connection().localAddress()->ip()->version();
+
+  if (write_callbacks_->connection().streamInfo().filterState()->hasData<TcpProxy::DownstreamAddrs>(
+          TcpProxy::DownstreamAddrs::key())) {
+    const auto downAddrs =
+        write_callbacks_->connection()
+            .streamInfo()
+            .filterState()
+            ->getDataReadOnly<TcpProxy::DownstreamAddrs>(TcpProxy::DownstreamAddrs::key());
+    ip_version = downAddrs.version();
+    src_address = downAddrs.srcAddress();
+    dst_address = downAddrs.dstAddress();
+    src_port = downAddrs.srcPort();
+    dstPort = downAddrs.dstPort();
+  }
+
+  switch (ip_version) {
+  case Network::Address::IpVersion::v4:
+    stream << PROXY_PROTO_V1_AF_INET << " ";
+    break;
+  case Network::Address::IpVersion::v6:
+    stream << PROXY_PROTO_V1_AF_INET6 << " ";
+    break;
+  }
+
+  stream << src_address << " ";
+  stream << dst_address << " ";
+  stream << src_port << " ";
+  stream << dstPort << "\r\n";
+
+  Buffer::OwnedImpl buff(stream.str());
+  write_callbacks_->injectWriteDataToFilterChain(buff, false);
+}
+
+void Filter::injectHeaderV2() {
+  Buffer::OwnedImpl buff(PROXY_PROTO_V2_SIGNATURE, PROXY_PROTO_V2_SIGNATURE_LEN);
+
+  if (!write_callbacks_->connection()
+           .streamInfo()
+           .filterState()
+           ->hasData<TcpProxy::DownstreamAddrs>(TcpProxy::DownstreamAddrs::key())) {
```

Comment:
Sorry if I'm missing something obvious, but why are we using this by default here and not in V1?

Reply:
In V2 of the protocol there's indication of whether the connection is being proxied or was intentionally generated locally by the proxy server. In the local case, no address information is sent. This check checks if the `DownstreamAddrs` info isn't available, and then just defaults to sending the local connection indicator. 

V1 also defaults to local, but I just coded it in a slightly different way since in V1 you do send the address info, even if it's a local connection.